qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -0,0 +1,324 @@
+
+    Args:
+        equiv_lib (EquivalenceLibrary): Source of valid translations
+        source_basis (Set[Tuple[gate_name: str, gate_num_qubits: int]]): Starting basis.
+        target_basis (Set[gate_name: str]): Target basis.
+        heuristic (Callable[[source_basis, target_basis], int]): distance heuristic.
+
+    Returns:
+        Optional[List[Tuple[gate, equiv_params, equiv_circuit]]]: List of (gate,
+            equiv_params, equiv_circuit) tuples tuples which, if applied in order
+            will map from source_basis to target_basis. Returns None if no path
+            was found.
+    """
+
+    source_basis = frozenset(source_basis)
+    target_basis = frozenset(target_basis)
+
+    open_set = set()  # Bases found but not yet inspected.
+    closed_set = set()  # Bases found and inspected.
+
+    # Priority queue for inspection order of open_set. Contains Tuple[priority, count, basis]
+    open_heap = []
+
+    # Map from bases in closed_set to predecessor with lowest cost_from_source.
+    # Values are Tuple[prev_basis, gate_name, params, circuit].
+    came_from = {}
+
+    basis_count = iter_count()  # Used to break ties in priority.
+
+    open_set.add(source_basis)
+    heappush(open_heap, (0, next(basis_count), source_basis))
+
+    # Map from basis to lowest found cost from source.
+    cost_from_source = defaultdict(lambda: np.inf)
+    cost_from_source[source_basis] = 0
+
+    # Map from basis to cost_from_source + heuristic.
+    est_total_cost = defaultdict(lambda: np.inf)
+    est_total_cost[source_basis] = heuristic(source_basis, target_basis)
+
+    logger.debug('Begining basis search from %s to %s.',
+                 source_basis, target_basis)
+
+    while open_set:
+        _, _, current_basis = heappop(open_heap)
+
+        if current_basis in closed_set:
+            # When we close a node, we don't remove it from the heap,
+            # so skip here.
+            continue
+
+        if set(gate_name for gate_name, gate_num_qubits in current_basis).issubset(target_basis):
+            # Found target basis. Construct transform path.
+            rtn = []
+            last_basis = current_basis
+            while last_basis != source_basis:
+                prev_basis, gate_name, gate_num_qubits, params, equiv = came_from[last_basis]
+
+                rtn.append((gate_name, gate_num_qubits, params, equiv))
+                last_basis = prev_basis
+            rtn.reverse()
+
+            logger.debug('Transformation path:')
+            for gate_name, gate_num_qubits, params, equiv in rtn:
+                logger.debug('%s/%s => %s\n%s', gate_name, gate_num_qubits, params, equiv)
+            return rtn
+
+        logger.debug('Inspecting basis %s.', current_basis)
+        open_set.remove(current_basis)
+        closed_set.add(current_basis)
+
+        for gate_name, gate_num_qubits in current_basis:
+            equivs = equiv_lib._get_equivalences((gate_name, gate_num_qubits))
```

Comment:
This is roughly the same as `EquivalenceLibrary.get_entry`, but called with a `(name, num_qubits)` tuple instead of a `Gate` instance (though I agree it's weird to use a private method here).

Reply:
so it is possible to do create a `Gate` instance out of `name` and `num_qubits` then look up based on that?

Reply:
Possibly, but we'd need a way to know the right `num_params` (unless we also tracked that in the basis nodes). I think the right next step is to move the basis nodes from `name/num_qubits` to something more like a `Gate` subclass.